File: client/src/nvim/notify.rs
use super::NvimGlobal;
use mlua::prelude::*;
use xbase_proto::MessageLevel;

pub trait NvimNotify {
    fn notify<S: AsRef<str>>(&self, msg: S, level: MessageLevel) -> LuaResult<()>;
}

impl NvimNotify for Lua {
    fn notify<S: AsRef<str>>(&self, msg: S, level: MessageLevel) -> LuaResult<()> {
        if msg.as_ref().trim().is_empty() {
            return Ok(());
        }

        let msg = msg.as_ref();
        let vim = self.vim()?;
        let level = level as u8;
        let opts = self.create_table_from([("title", "XBase")])?;
        let args = (msg, level, opts);

        // NOTE: Plugins like nvim-notify sets notify to metatable
        match vim.get::<_, LuaFunction>("notify") {
            Ok(f) => f.call(args),
            Err(_) => vim.get::<_, LuaTable>("notify")?.call(args),
        }
    }
}

File: lua/xbase/notify.lua
local M = {}
local notify = function(msg, level)
  if #vim.trim(msg) == 0 then
    return
  end
  if level == 5 then
    vim.api.nvim_echo({ { msg, "healthSuccess" } }, true, {})
    return
  end

  vim.notify(msg, level, {
    title = "XBase",
  })
end

M = setmetatable(M, {
  __index = function(_, level)
    level = vim.log.levels[string.upper(level)]
    return function(msg)
      return notify(msg, level)
    end
  end,
  __call = function(_, msg, level)
    return notify(msg, vim.log.levels[level:upper()] or 5)
  end,
})

return M
